### ISSUE TITLE
Rule proposal: Disallow variables declared with `[await] using` from being passed beyond the lifetime of their declaration scope

### ISSUE THREAD

**Issue by Renegade334** [typescript-eslint/typescript-eslint#11309]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Explicit resource management is a new language paradigm, and carries with it potential footguns.

Most notably, there's nothing to stop variables declared with `[await] using` having their contents passed outside of the scope in which they're declared, in such a way that the disposable resource persists beyond the lifetime of the declaring scope. This is almost invariably unintentional, as the resource in question will be disposed of when execution leaves the original scope.

```ts
function getDisposableResource() {
  using resource = { ... }
  // ...
  return resource
} // oops, the resource is disposed upon exiting the function
```

For a variable `disposable` declared using `[await] using`, it should be possible for a rule to flag the following:
- assignment of the value of `disposable` to a variable outside the declaration scope
- returning `disposable` from a function
- exporting `disposable`
- ~~referencing `disposable` within a closure~~ (removed: https://github.com/typescript-eslint/typescript-eslint/issues/11309#issuecomment-2990358185)

### Fail Cases

```TypeScript
function makeMeADisposable() {
  using disposable = { ... }
  // ...
  return disposable
}

using disposable = { ... }
// ...
export { disposable }

let resource
for (using disposable of iterator) {
  if (condition) {
    resource = disposable
    break
  }
}
```

### Pass Cases

```TypeScript
function makeMeADisposable() {
  const disposable = { ... }
  // ...
  return disposable
}

const disposable = { ... }
// ...
export { disposable }

let property
for (using disposable of iterator) {
  if (condition) {
    property = disposable.property
    break
  }
}
```

### Additional Info

`export using disposable = ...` is an illegal pattern under the specification, but `export { disposable }` is not.

**Comment by kirkwaiblinger** [comment]
Related: https://github.com/typescript-eslint/typescript-eslint/issues/8255

**Comment by JoshuaKGoldberg** [comment]
I really like the idea here. But a lot of the patterns probably fall into the same _"we can't know whether this is safe"_ limitations as https://typescript-eslint.io/rules/no-floating-promises.

Looking at the first failure case:

```ts
function makeMeADisposable() {
  using disposable = ...
  someSetupOperation(disposable)
  return disposable
}

function someSetupOperation(disposable) {
  console.log(disposable);
}
```

That code seems very fine and reasonable. It'd be really annoying as a user to have this reported on.

Not saying this is a wontfix. Just, I think for typescript-eslint we are limited to constructs that are _definitely_ unsafe - not just _probably under certain conditions_.

**Comment by Renegade334** [comment]
> Looking at the first failure case:
> 
> function makeMeADisposable() {
>   using disposable = ...
>   someSetupOperation(disposable)
>   return disposable
> }
> 
> function someSetupOperation(disposable) {
>   console.log(disposable);
> }
> 
> That code seems very fine and reasonable. It'd be really annoying as a user to have this reported on.

Sorry, should have been more specific. Passing a disposable object around is absolutely fine, providing the declaration scope is still alive. It's the `return` which is problematic – since by definition, the object will have its disposer invoked after the `return`, so whatever is receiving the object from the function call is going to be receiving it in a disposed state.

These patterns are chosen to be the ones where the disposable object is guaranteed to be accessible beyond its original declaration scope, ergo beyond the point at which it has been disposed. The one which is less certain is the closure example, depending on how exactly the closure ends up being used.

**Comment by kirkwaiblinger** [comment]
I think I'm +1 to a rule that would flag these kinds of cases. Maybe "no-use-after-dispose"? To parallel "no-use-before-define".

Could be worth perusing https://github.com/tc39/proposal-explicit-resource-management (for example, discussion in https://github.com/tc39/proposal-explicit-resource-management/issues/244) for other related footguns? 

For now, my thought is

```ts
// definitely bad, should flag
function foo() {
  using x = /* ... */;
  return x; // and friends `throw`, `yield`
}

// definitely bad, should flag
using x = /* (at top level of module) */
export { x }

// can't flag because we can't really generalize about closures
{
  using disposable = ...
  someCallbackStyleAsyncFunction((err, data) => {
    doSomething(disposable, data) // closure invoked after disposal
  })
}
// counterexample - no problem here
function foo() {
   using x = /* ... */, y = /* ... */;
   const isSomethingTrue = [x, y].some(z => z.something()); // closure invoked before disposal
}

// on the fence about
let resource
for (using disposable of iterator) {
  if (condition) {
    resource = disposable
    break
  }
}
// because variables in outer scopes may be declared that way without the implication 
// that they'll be _used_ in the outer scope...
let resource
for (using disposable of iterator) {
  resource = null;
  if (condition) {
    resource = disposable
  } 
  resource?.someMethod();
}
// ...but it's also usually easy to rescope the variable in that case...
for (using disposable of iterator) {
  let resource = null;
  if (condition) {
    resource = disposable
  } 
  resource?.someMethod();
}
// ... so more examples may be needed in order to sus this one out.
```

**Comment by Renegade334** [comment]
> ```ts
> // can't flag because we can't really generalize about closures
> {
>   using disposable = ...
>   someCallbackStyleAsyncFunction((err, data) => {
>     doSomething(disposable, data) // closure invoked after disposal
>   })
> }
> // counterexample - no problem here
> function foo() {
>    using x = /* ... */, y = /* ... */;
>    const isSomethingTrue = [x, y].some(z => z.something()); // closure invoked before disposal
> }
> ```

+1 – this was motivated by a real-world case, but it could only really be an optional configurable at most for this reason, and easier to omit entirely.

> ```ts
> // on the fence about
> let resource
> for (using disposable of iterator) {
>   if (condition) {
>     resource = disposable
>     break
>   }
> }
> // because variables in outer scopes may be declared that way without the implication 
> // that they'll be _used_ in the outer scope...
> let resource
> for (using disposable of iterator) {
>   resource = null;
>   if (condition) {
>     resource = disposable
>   } 
>   resource?.someMethod();
> }
> // ...but it's also usually easy to rescope the variable in that case...
> for (using disposable of iterator) {
>   let resource = null;
>   if (condition) {
>     resource = disposable
>   } 
>   resource?.someMethod();
> }
> // ... so more examples may be needed in order to sus this one out.
> ```

The behaviour of `using` variables is so intrinsically tied to their declaring scope that I think this is legitimate to flag. Although eslint can't infer how the super-scoped variable is being used in practice, it can be confident that the disposable will have a live reference beyond the point of disposal (unless the super-scoped variable is cleared after use, but then there's no reason not to use a same-scope variable instead).

**Comment by kirkwaiblinger** [comment]
@typescript-eslint/triage-team 

Opening this to PRs, since, at a bare minimum, flagging `return usingVariable` seems like a no-brainer to lint for, and I think the other proposed checks are well motivated as well. But yell if there's an objection!

**Comment by kirkwaiblinger** [comment]
Hang on, actually, sorry - does this check require type information or deep TS knowledge in any way? I don't think so, since it seems like these proposed checks are just normal scope analysis right? If it does, we can take this, but, if not, now that eslint core is supporting `using`/`await using` (https://github.com/eslint/eslint/issues/19792) this should be proposed there first, actually.

**Comment by Renegade334** [comment]
You're right, it should only need to observe the ESTree. My original brainstorming did involve some type observation, but this list of cases doesn't.

**Comment by bradzacher** [comment]
This is also hard because for example this code is also 100% valid code:
```ts
interface DbConnection extends Disposable {
  lastResults: unknown;
  query(sql: string): void;
}
declare function getConn(): DbConnection;

export function getData(): { lastResults: unknown } {
  using conn = getConn();
  conn.query('SELECT * FROM lint_rules');
  return conn;
}
```

Just because something is disposable doesn't mean that, after it's disposed:
a) it isn't still valid to use the thing in some limited capacity
b) (following on from that) the thing can't be type-narrowed to just its "safe" members and passed around.

I think there is scope for a rule here, but I just don't know if there's enough weight in the ecosystem to definitively declare something as a "good" or "bad" pattern just yet. As we are such a fundamental part of the ecosystem -- us making a rule kind of declares something as exactly that, and I don't know if we should be making such decisions about the "blessed" ways to use new *_JS_* features.

---

As Kirk pointed out -- there's nothing here that requires type info and there's nothing here that is TS-specific -- so I think we should reject this as a proposal for typescript-eslint.

**Comment by Renegade334** [comment]
Thought this was already closed!

Thanks for the discussion – I'll look to propose this across the way.
